xc/common/cmake/vivado.cmake

```diff
@@ -464,3 +478,199 @@ function(ADD_VIVADO_PNR_TARGET)
       DEPS ${DEPS}
       BITSTREAM ${BITSTREAM})
 endfunction()
+
+function(PREPARE_RAPIDWRIGHT)
+  # Creates a target rapidwright with the following envirnoment variables:
+  #  RAPIDWRIGHT_INSTALLED - True if a valid RapidWright installation is
+  #                          found at the variable RAPIDWRIGHT_PATH.
+  #  RAPIDWRIGHT_PATH - If RAPIDWRIGHT_INSTALLED=TRUE, this points to the
+  #                     directory where RapidWright is located.
+  #  JAVA - If RAPIDWRIGHT_INSTALLED=TRUE, this points to the location of
+  #         the Java runtime executable needed invoke RapidWright entry
+  #         points.
+  #  INVOKE_RAPIDWRIGHT - If RAPIDWRIGHT_INSTALLED=TRUE, this points to the
+  #         location of the invoke_rapidwright.sh script which can be used to
+  #         invoke RapidWright.
+  add_custom_target(rapidwright)
+  set_target_properties(rapidwright PROPERTIES RAPIDWRIGHT_INSTALLED FALSE)
+
+  set(HAVE_RAPIDWRIGHT FALSE)
+  set(HAVE_JAVA FALSE)
+  set(HAVE_URAY_VIVADO FALSE)
+  if(EXISTS "${RAPIDWRIGHT_PATH}" AND EXISTS ${RAPIDWRIGHT_PATH}/interchange AND EXISTS ${RAPIDWRIGHT_PATH}/scripts/invoke_rapidwright.sh)
+    set(HAVE_RAPIDWRIGHT TRUE)
+  endif()
+
+  if(${HAVE_RAPIDWRIGHT})
+    find_program(JAVA java)
+    if(EXISTS ${JAVA})
+      set(HAVE_JAVA TRUE)
+    endif()
+  endif()
+
+  if(DEFINED ENV{URAY_VIVADO_SETTINGS} AND EXISTS "$ENV{URAY_VIVADO_SETTINGS}")
```

Comment:
How `URAY_VIVADO_SETTINGS` is U-Ray specific? Is the same Vivado is used for xc7 targets?

Reply:
RapidWright needs a more recent version of Vivado, and URAY_VIVADO_SETTINGS currently conforms to the minimum Vivado requirement.